File: src/entities/address.cpp
// Copyright (c) 2019-2020 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

#include "veriblock/entities/address.hpp"

#include "veriblock/assert.hpp"
#include "veriblock/base58.hpp"
#include "veriblock/base59.hpp"
#include "veriblock/consts.hpp"
#include "veriblock/hashutil.hpp"
#include "veriblock/serde.hpp"

namespace altintegration {

constexpr const auto STARTING_CHAR = 'V';
constexpr const auto MULTISIG_ENDING_CHAR = '0';

constexpr const auto MULTISIG_ADDRESS_M_VALUE = 1;
constexpr const auto MULTISIG_ADDRESS_N_VALUE = 2;
constexpr const auto MULTISIG_ADDRESS_MIN_N_VALUE = 2;
constexpr const auto MULTISIG_ADDRESS_MAX_N_VALUE = 58;
constexpr const auto MULTISIG_ADDRESS_MAX_M_VALUE = 58;
constexpr const auto MULTISIG_ADDRESS_DATA_END = 24;
constexpr const auto MULTISIG_ADDRESS_CHECKSUM_END = 28;

static std::string getDataPortionFromAddress(const std::string& address) {
  VBK_ASSERT(address.length() == ADDRESS_SIZE);
  return address.substr(0, MULTISIG_ADDRESS_DATA_END + 1);
}

static bool isMultisig(const std::string& address) {
  VBK_ASSERT(address.length() == ADDRESS_SIZE);
  return (address[ADDRESS_SIZE - 1] == MULTISIG_ENDING_CHAR);
}

static std::string getChecksumPortionFromAddress(const std::string& address,
                                                 bool multisig) {
  VBK_ASSERT(address.length() == ADDRESS_SIZE);
  if (multisig) {
    return address.substr(
        MULTISIG_ADDRESS_DATA_END + 1,
        MULTISIG_ADDRESS_CHECKSUM_END - MULTISIG_ADDRESS_DATA_END);
  }
  return address.substr(MULTISIG_ADDRESS_DATA_END + 1);
}

static bool isBase58String(const std::string& input) {
  std::vector<uint8_t> out;
  return DecodeBase58(input, out);
}

static bool isBase59String(const std::string& input) {
  std::vector<uint8_t> out;
  return DecodeBase59(input, out);
}

static std::string calculateChecksum(const std::string& data, bool multisig) {
  auto hash = sha256(data);
  std::string checksum = EncodeBase58(hash);
  if (multisig) {
    return checksum.substr(0, 3 + 1);
  }
  return checksum.substr(0, 4 + 1);
}

static std::string addressChecksum(const Address& address) {
  return calculateChecksum(address.toString(),
                           address.getType() == AddressType::MULTISIG);
}

Address Address::fromPublicKey(Slice<const uint8_t> publicKey) {
  auto keyHash = sha256(publicKey);
  auto keyHashEncoded = EncodeBase58(keyHash);
  auto data = std::string{STARTING_CHAR} +
              keyHashEncoded.substr(0, MULTISIG_ADDRESS_DATA_END);
  auto checksum = calculateChecksum(data, false);
  return Address(AddressType::STANDARD, data + checksum);
}

bool Address::isDerivedFromPublicKey(Slice<const uint8_t> publicKey) const {
  auto expectedAddress = fromPublicKey(publicKey);
  if (*this != expectedAddress) {
    return false;
  }
  if (addressChecksum(*this) != addressChecksum(expectedAddress)) {
    return false;
  }
  return true;
}

bool Address::fromString(const std::string& input, ValidationState& state) {
  if (input.size() != ADDRESS_SIZE) {
    return state.Invalid(
        "addr-bad-length",
        fmt::format("Invalid address length. Expected={}, got={}.",
                    ADDRESS_SIZE,
                    input.size()));
  }

  if (input[0] != STARTING_CHAR) {
    return state.Invalid(
        "addr-bad-starting-char",
        fmt::format("Address should start with {}", STARTING_CHAR));
  }

  std::string data = getDataPortionFromAddress(input);
  bool multisig = isMultisig(input);
  std::string checksum = getChecksumPortionFromAddress(input, multisig);

  if (multisig) {
    if (!isBase59String(input)) {
      return state.Invalid("addr-bad-multisig-content", "Not a base59 string.");
    }

    /* To make the addresses 'human-readable' we add 1 to the decoded value (1
     * in Base58 is 0, but we want an address with a '1' in the m slot to
     * represent m=1, for example). this allows addresses with m and n both <=
     * 9 to be easily recognized. Additionally, an m or n value of 0 makes no
     * sense, so this allows multisig to range from 1 to 58, rather than what
     * would have otherwise been 0 to 57. */
    auto decodeNumber = [](const std::string& in, int& num) -> bool {
      std::vector<uint8_t> ret(in.size(), 0);
      if (!DecodeBase58(in, ret)) {
        return false;
      }
      num = ret[0] + 1;
      return true;
    };

    int n = 0, m = 0;
    if (!decodeNumber(std::string(1, input[MULTISIG_ADDRESS_M_VALUE]), m)) {
      return state.Invalid("addr-multisig-bad-m");
    }
    if (!decodeNumber(std::string(1, input[MULTISIG_ADDRESS_N_VALUE]), n)) {
      return state.Invalid("addr-multisig-bad-n");
    }

    if (n < MULTISIG_ADDRESS_MIN_N_VALUE) {
      return state.Invalid(
          "addr-multisig-bad-n-too-small",
          fmt::format("Expected N to be at least {}, but got {}",
                      MULTISIG_ADDRESS_MIN_N_VALUE,
                      n));
    }
    if (m > n) {
      return state.Invalid("addr-multisig-bad-mn",
                           "Address has more signatures than addresses");
    }
    if ((n > MULTISIG_ADDRESS_MAX_N_VALUE) ||
        (m > MULTISIG_ADDRESS_MAX_M_VALUE)) {
      return state.Invalid("addr-multisig-bad-mn-too-many",
                           "Too many addresses or signatures");
    }

    if (!isBase58String(input.substr(0, ADDRESS_SIZE - 1))) {
      return state.Invalid("addr-multisig-bad-remainder",
                           "Remainder is not a base58 string");
    }
  } else {
    if (!isBase58String(input)) {
      return state.Invalid("addr-bad", "Address is not a base58 string");
    }
  }

  auto expectedChecksum = calculateChecksum(data, multisig);
  if (expectedChecksum != checksum) {
    return state.Invalid(
        "addr-bad-checksum",
        fmt::format("Checksum does not match. Expected={}, got={}.",
                    expectedChecksum,
                    checksum));
  }

  m_Type = multisig ? AddressType::MULTISIG : AddressType::STANDARD;
  m_Address = input;

  return true;
}

std::string Address::toString() const noexcept { return m_Address; }

void Address::toVbkEncoding(WriteStream& stream) const {
  stream.writeBE<uint8_t>((uint8_t)getType());
  std::vector<uint8_t> decoded;
  switch (getType()) {
    case AddressType::STANDARD:
      decoded = AssertDecodeBase58(toString());
      break;
    case AddressType ::MULTISIG:
      decoded = AssertDecodeBase59(toString());
      break;
    default:
      // if we don't know address type, do not encode anything
      return;
  }

  writeSingleByteLenValue(stream, decoded);
}

size_t Address::estimateSize() const {
  size_t size = 0;
  size += sizeof((uint8_t)getType());
  std::vector<uint8_t> decoded;
  switch (getType()) {
    case AddressType::STANDARD:
      decoded = AssertDecodeBase58(toString());
      break;
    case AddressType ::MULTISIG:
      decoded = AssertDecodeBase58(toString());
      break;
    default:
      // if we don't know address type, do not encode anything
      return size;
  }

  size += singleByteLenValueSize(decoded);
  return size;
}

void Address::getPopBytes(WriteStream& stream) const {
  std::vector<uint8_t> bytes = AssertDecodeBase58(m_Address.substr(1));
  stream.write(std::vector<uint8_t>(
      bytes.begin(), bytes.begin() + ADDRESS_POP_DATA_SIZE_PROGPOW));
}

bool Address::operator==(const Address& other) const noexcept {
  return m_Address == other.m_Address;
}

bool Address::operator!=(const Address& other) const noexcept {
  return !(this->operator==(other));
}
bool Address::operator==(const std::string& other) const noexcept {
  return m_Address == other;
}

bool DeserializeFromVbkEncoding(ReadStream& stream,
                                Address& out,
                                ValidationState& state) {
  uint8_t addressType = 0;
  if (!stream.readBE<uint8_t>(addressType, state)) {
    return state.Invalid("address-type");
  }

  Slice<const uint8_t> addressBytes;
  if (!readSingleByteLenValue(
          stream, addressBytes, state, 0, altintegration::ADDRESS_SIZE)) {
    return state.Invalid("address-bytes");
  }

  std::string addressText;
  switch ((AddressType)addressType) {
    case AddressType::STANDARD:
      addressText = EncodeBase58(addressBytes);
      break;
    case AddressType::MULTISIG:
      addressText = EncodeBase59(addressBytes);
      break;
    default:
      return state.Invalid("invalid-address-type");
  }

  if (!out.fromString(addressText, state)) {
    return state.Invalid("bad-addr");
  }

  return true;
}

Address Address::assertFromString(const std::string& input) {
  Address addr;
  ValidationState state;
  bool success = addr.fromString(input, state);
  VBK_ASSERT_MSG(success, state.toString());
  return addr;
}

// this code verifies that default address is valid
// NOLINTNEXTLINE(cert-err58-cpp)
static Address defaultAddress = [] {
  Address addr;
  ValidationState state;
  bool success = addr.fromString("V111111111111111111111111G3LuZ", state);
  VBK_ASSERT_MSG(success, state.toString());
  return addr;
}();

Address::Address() {
  this->m_Address = defaultAddress.m_Address;
  this->m_Type = defaultAddress.m_Type;
}

}  // namespace altintegration

File: bindings/go/entities/address.go
package entities

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"io"

	veriblock "github.com/VeriBlock/alt-integration-cpp/bindings/go"
)

// AddressType ...
type AddressType uint8

const (
	// AddressTypeZeroUnused ..
	AddressTypeZeroUnused AddressType = iota
	// AddressTypeStandard ..
	AddressTypeStandard
	// AddressTypeProofOfProof ..
	AddressTypeProofOfProof
	// AddressTypeMultisig ..
	AddressTypeMultisig
)

// Consts
const (
	StartingChar       = 'V'
	MultisigEndingChar = '0'

	MultisigAddressMValue      = 1
	MultisigAddressNValue      = 2
	MultisigAddressMinNValue   = 2
	MultisigAddressMaxNValue   = 58
	MultisigAddressMaxMValue   = 58
	MultisigAddressDataEnd     = 24
	MultisigAddressChecksumEnd = 28
)

// Address ...
type Address struct {
	addressType AddressType
	address     string
}

func getDataPortionFromAddress(address string) (string, error) {
	if len(address) != veriblock.AddressSize {
		return "", errors.New("Invalid address size")
	}
	return address[0 : MultisigAddressDataEnd+1], nil
}

func isMultisig(address string) (bool, error) {
	if len(address) != veriblock.AddressSize {
		return false, errors.New("Invalid address size")
	}
	return (address[veriblock.AddressSize-1] == MultisigEndingChar), nil
}

func getChecksumPortionFromAddress(address string, multisig bool) (string, error) {
	if len(address) != veriblock.AddressSize {
		return "", errors.New("Invalid address size")
	}
	if multisig {
		return address[MultisigAddressDataEnd+1 : MultisigAddressChecksumEnd+1], nil
	}
	return address[MultisigAddressDataEnd+1:], nil
}

func isBase58String(input string) bool {
	_, err := veriblock.DecodeBase58(input)
	return err == nil
}

func isBase59String(input string) bool {
	_, err := veriblock.DecodeBase59(input)
	return err == nil
}

func calculateChecksum(data string, multisig bool) string {
	hash := sha256.Sum256([]byte(data))
	checksum := veriblock.EncodeBase58(hash[:])
	if multisig {
		return checksum[0 : 3+1]
	}
	return checksum[0 : 4+1]
}

func addressChecksum(address Address) string {
	return calculateChecksum(address.ToString(), address.GetType() == AddressTypeMultisig)
}

// FromPublicKey ...
func (v *Address) FromPublicKey(publicKey []byte) {
	keyHash := sha256.Sum256(publicKey)
	keyHashEncoded := veriblock.EncodeBase58(keyHash[:])
	data := "V" + keyHashEncoded[0:MultisigAddressDataEnd]
	checksum := calculateChecksum(data, false)
	*v = Address{AddressTypeStandard, data + checksum}
}

// IsDerivedFromPublicKey ...
func (v *Address) IsDerivedFromPublicKey(publicKey []byte) bool {
	expectedAddress := Address{}
	expectedAddress.FromPublicKey(publicKey)
	if *v != expectedAddress {
		return false
	}
	if addressChecksum(*v) != addressChecksum(expectedAddress) {
		return false
	}
	return true
}

// FromVbkEncoding ...
func (v *Address) FromVbkEncoding(stream io.Reader) error {
	var addressType uint8
	if err := binary.Read(stream, binary.LittleEndian, &addressType); err != nil {
		return err
	}
	addressBytes, err := veriblock.ReadSingleByteLenValue(stream, 0, veriblock.AddressSize)
	if err != nil {
		return err
	}
	var addressText string
	switch AddressType(addressType) {
	case AddressTypeStandard:
		addressText = veriblock.EncodeBase58(addressBytes)
		break
	case AddressTypeMultisig:
		addressText = veriblock.EncodeBase59(addressBytes)
		break
	default:
		return errors.New("invalid-address-type")
	}
	return v.FromString(addressText)
}

// FromVbkEncodingBytes ...
func (v *Address) FromVbkEncodingBytes(data []byte) error {
	buffer := bytes.NewBuffer(data)
	return v.FromVbkEncoding(buffer)
}

// FromString ...
func (v *Address) FromString(str string) error {
	address, err := NewAddress(str)
	if err != nil {
		return err
	}
	*v = *address
	return nil
}

// ToVbkEncoding ...
func (v *Address) ToVbkEncoding(stream io.Writer) error {
	binary.Write(stream, binary.BigEndian, byte(v.GetType()))
	var decoded []byte
	var err error
	switch v.GetType() {
	case AddressTypeStandard:
		decoded, err = veriblock.DecodeBase58(v.ToString())
		break
	case AddressTypeMultisig:
		decoded, err = veriblock.DecodeBase59(v.ToString())
		break
	default:
		// If we don't know address type, do not encode anything
		return errors.New("Unsupported address type")
	}
	if err != nil {
		return err
	}
	veriblock.WriteSingleByteLenValue(stream, decoded)
	return nil
}

// ToVbkEncodingBytes ...
func (v *Address) ToVbkEncodingBytes() ([]byte, error) {
	var buffer bytes.Buffer
	err := v.ToVbkEncoding(&buffer)
	return buffer.Bytes(), err
}

// ToString ...
func (v *Address) ToString() string {
	return v.address
}

// GetType ...
func (v *Address) GetType() AddressType {
	return v.addressType
}

// NewAddress ...
func NewAddress(input string) (*Address, error) {
	if len(input) != veriblock.AddressSize {
		return nil, errors.New("isValidAddress(): invalid address length")
	}
	if input[0] != StartingChar {
		return nil, errors.New("isValidAddress(): not a valid VBK address")
	}
	data, err := getDataPortionFromAddress(input)
	if err != nil {
		return nil, err
	}
	multisig, err := isMultisig(input)
	if err != nil {
		return nil, err
	}
	checksum, err := getChecksumPortionFromAddress(input, multisig)
	if err != nil {
		return nil, err
	}
	if multisig {
		if !isBase59String(input) {
			return nil, errors.New("isValidAddress(): not a base59 string")
		}
		/* To make the addresses 'human-readable' we add 1 to the decoded value (1
		 * in Base58 is 0, but we want an address with a '1' in the m slot to
		 * represent m=1, for example). this allows addresses with m and n both <=
		 * 9 to be easily recognized. Additionally, an m or n value of 0 makes no
		 * sense, so this allows multisig to range from 1 to 58, rather than what
		 * would have otherwise been 0 to 57. */
		decodedM, err := veriblock.DecodeBase58(string(input[MultisigAddressMValue]))
		decodedN, err := veriblock.DecodeBase58(string(input[MultisigAddressNValue]))
		if err != nil {
			return nil, err
		}
		n := decodedN[0] + 1
		m := decodedM[0] + 1
		if n < MultisigAddressMinNValue {
			return nil, errors.New("isValidAddress(): not enough addresses to be multisig")
		}
		if m > n {
			return nil, errors.New("isValidAddress(): address has more signatures than addresses")
		}
		if (n > MultisigAddressMaxNValue) ||
			(m > MultisigAddressMaxMValue) {
			return nil, errors.New("isValidAddress(): too many addresses/signatures")
		}
		if !isBase58String(input[0 : veriblock.AddressSize-1]) {
			return nil, errors.New("isValidAddress(): remainder is not a base58 string")
		}
	} else {
		if !isBase58String(input) {
			return nil, errors.New("isValidAddress(): address is not a base58 string")
		}
	}
	expectedChecksum := calculateChecksum(data, multisig)
	if expectedChecksum != checksum {
		return nil, errors.New("isValidAddress(): checksum does not match")
	}
	addressType := AddressTypeStandard
	if multisig {
		addressType = AddressTypeMultisig
	}
	return &Address{addressType, input}, nil
}
